### PR TITLE
[clang] Fix loss of `dllexport` for exported template specialization

### PR DESCRIPTION
**Pull request by nga888**
In commit 0a20f541, "Better codegen support for DLL attributes being dropped after the first declaration (PR20792)", code was added to enable "dropping" of DLL attributes. The specific issue and example given was related to `dllimport` and this was the test case that was added in that commit.

However, the code also included the "dropping" of `dllexport` but no test case for this was added. This "dropping" of `dllexport` can cause a `dllexport` template specialization to incorrectly lose its `dllexport` attribute as shown by the test case in this patch.

As there appears to be no need for the "dropping" of `dllexport`, remove this code to fix this issue.

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 1]
```diff
@@ -4849,9 +4852,12 @@ CodeGenModule::GetOrCreateLLVMGlobal(StringRef MangledName, llvm::Type *Ty,
...
 
-    // Handle dropped DLL attributes.
-    if (D && !D->hasAttr<DLLImportAttr>() && !D->hasAttr<DLLExportAttr>() &&
-        !shouldMapVisibilityToDLLExport(D))
+    // Handle dropped dllimport.
+    if (D &&
+        (Entry->getDLLStorageClass() ==
+         llvm::GlobalVariable::DLLImportStorageClass) &&
+        !D->hasAttr<DLLImportAttr>() &&
+        !shouldMapVisibilityToDLLExport(cast_or_null<NamedDecl>(D)))
```
Comment (pogo59) [inline clang/lib/CodeGen/CodeGenModule.cpp:4860]:
```suggestion
        !shouldMapVisibilityToDLLExport(cast<NamedDecl>(D)))
```
D is non-null here.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 2]
```diff
@@ -4554,8 +4554,11 @@ llvm::Constant *CodeGenModule::GetOrCreateLLVMFunction(
...
     }
 
-    // Handle dropped DLL attributes.
-    if (D && !D->hasAttr<DLLImportAttr>() && !D->hasAttr<DLLExportAttr>() &&
+    // Handle dropped dllimport.
+    if (D &&
+        (Entry->getDLLStorageClass() ==
+         llvm::GlobalVariable::DLLImportStorageClass) &&
+        !D->hasAttr<DLLImportAttr>() &&
         !shouldMapVisibilityToDLLExport(cast_or_null<NamedDecl>(D))) {
```
Comment (pogo59) [inline clang/lib/CodeGen/CodeGenModule.cpp:4562]:
```suggestion
        !shouldMapVisibilityToDLLExport(cast<NamedDecl>(D))) {
```
D is non-null here.

**Review by pogo59 — COMMENTED** [review]
Does the test exercise both modified paths? I'm guessing it only exercises GetOrCreateLLVMFunction, but I'm not a codegen expert.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 1 above)
Comment (nga888) [reply clang/lib/CodeGen/CodeGenModule.cpp:4860]: This part of the code isn't directly related to the change but I assume I can go ahead and change it?

**Comment by nga888** [comment]
> Does the test exercise both modified paths? I'm guessing it only exercises GetOrCreateLLVMFunction, but I'm not a codegen expert.

Only 1 path is "exercised" but I believe neither path of the removed code was exercised previously and no other tests are affected by the code removal.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 1 above)
Comment (nga888) [reply clang/lib/CodeGen/CodeGenModule.cpp:4860]: Sorry, the above comment should have been in reply to the instance above. This cast shouldn't be needed at all. I suspect a copy & paste error on my behalf!

Comment (nga888) [reply clang/lib/CodeGen/CodeGenModule.cpp]: I've fixed up both instances of the call to `shouldMapVisibilityToDLLExport()`.

**Review by zmodem — COMMENTED** [review]
> In commit https://github.com/llvm/llvm-project/commit/0a20f5417c9d241f4774a49da6c7ca8123686671, "Better codegen support for DLL attributes being dropped after the first declaration (PR20792)", code was added to enable "dropping" of DLL attributes.

Wow, how is that 10 years ago already?

I'm not sure this is the right place for the fix though. Note that even though my patch added handling for dropping dllexport, we still do export `f` in this case:

```
__declspec(dllexport) int f(int x);
int user(int x) {
  return f(x);
}
int f(int x) { return 1; }
```

Because `f` is still dllexport at the AST level, as it should. That's handled by the code which deals with redeclarations.

I suspect that's where we should fix your case also: we should figure out why `s::f<true>()` isn't(?) marked dllexport in the AST.

**Comment by zmodem** [comment]
> I understand that removing the code that drops `dllexport` may not be the "best" place to fix this issue but if this code is not actually required, which appears to be the case (please let me know if this is not), then removing the code to achieve "better" behaviour feels like a "win". The alternative would be to add more code/complexity to the AST handling.

I still think fixing the AST would be better. That would also fix the "dllimport version" of your test case:

```
struct s {
  template <bool b = true> static bool f();
};
template <typename T> bool template_using_f(T) { return s::f(); }
bool use_template_using_f() { return template_using_f(0); }

template<> inline
bool __declspec(dllimport) s::f<true>() { return true; }
```

Here, `dllimport` gets dropped from `s::f<true>()` for the same reason as in your test: the AST node which the call in `template_using_f()` refers to is missing the attribute. (Godbolt: https://godbolt.org/z/nnfET4r13)

**Comment by nga888** [comment]
> Here, `dllimport` gets dropped from `s::f<true>()` for the same reason as in your test: the AST node which the call in `template_using_f()` refers to is missing the attribute. (Godbolt: https://godbolt.org/z/nnfET4r13)

I think the difference in this case, is that this behaviour is unlikely to cause a "real" issue. The import of a definition/specialization is already a bit of an anomaly, so dropping `dllimport` is probably reasonable.

**Review by zmodem — CHANGES_REQUESTED** [review]
> The import of a definition/specialization is already a bit of an anomaly, so dropping dllimport is probably reasonable.

Well, you're exporting a specialization. How is it going to get imported?

Perhaps me leaving the definition in there was unnecessary. How about just:

```
template<>
bool __declspec(dllimport) s::f<true>();
```
(https://godbolt.org/z/nqE7zPWYP)

Anyway, my point is that your patch is only working around half the bug. It seems to me that the proper fix has to be to get the attributes right in the AST.

**Comment by nga888** [comment]
> Well, you're exporting a specialization. How is it going to get imported?

The "importing" side is just:
```
struct s {
  template <bool b = true> static bool f();
};
bool use_f() { return s::f(); }
```
This test case was derived from code that is part of UnrealEngine 5.4 which builds fine with MSVC, i.e. no undefined symbol for `s::f<true>()`.

> Anyway, my point is that your patch is only working around half the bug. It seems to me that the proper fix has to be to get the attributes right in the AST.

So in this test case for exported template specializations, there isn't really another "half" to the issue.

**Comment by nga888** [comment]
> I think the "surprise" is that the latter call to `GetOrCreateLLVMFunction` is using an earlier `FunctionDecl`. Because of that, I'm thinking maybe the fix is to call `getMostRecentDecl` on it, something like:

Thanks for your time and the update. Yes, this is what I also discovered whilst debugging this issue. I also came to the same conclusion as you but when I tried using `getMostRecentDecl()` it didn't appear to work for me. Have you confirmed that this does fix this issue? I will look again, perhaps I made a mistake when trying this fix.

It was after this failure that I then started to question the whole removing of `dllexport`. I still think that this doesn't serve any useful purpose (and it isn't tested) which is why I would still favour the removal of the code.

**Comment by zmodem** [comment]
> Ok, I have confirmed that the getMostRecentDecl() fix does indeed work on the main branch.

Great!

> However, this fix also causes a dllimport test to fail too, so perhaps that was the "failure" that I encountered, I just don't remember.

Was this about `@_ZdlPv` in `clang/test/CodeGenCXX/dllimport.cpp`? Using `getMostRecentDecl()` makes that dllimport, and I think that's correct and matching MinGW's behavior: https://godbolt.org/z/Mzo1zGsrj

I think we should proceed with switching to `getMostRecentDecl()`. Either a new PR or continuing in this one works for me.

**Comment by nga888** [comment]
@zmodem, forgot to ask, and not directly linked to this patch as such, but shouldn't there be a test case added for the dropping of `dllexport`?

**Comment by zmodem** [comment]
> The odd thing is that this behaviour doesn't change on the main branch or at least not with the version of it that I tried on Friday.

I could have sworn it failed... but I was also using main from a few weeks ago. I think the test doesn't fail anymore because sized delete was enabled in 130e93cc26ca9d3ac50ec5a92e3109577ca2e702. Maybe we should specify -fno-sized-deallocation or update the test somehow.

> shouldn't there be a test case added for the dropping of dllexport?

I can't think of a test case that would hit it.

**Comment by nga888** [comment]
> > shouldn't there be a test case added for the dropping of dllexport?
> 
> I can't think of a test case that would hit it.

Just for clarification, you want to pursue the `getMostRecentDecl()` approach but keep the code for the dropping of `dllexport` even though there appears to be no test case that would/can exercise this code?

**Comment by nga888** [comment]
Closing this PR in favour of PR #94664.

**Pull request closed without merging**